htlcswitch/switch.go

```diff
@@ -367,9 +370,20 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID,
 		return zeroPreimage, err
 	}
 
+	paymentID, err := s.paymentSequencer.NextID()
+	if err != nil {
+		return zeroPreimage, err
+	}
+
+	// To ensure atomicity between checking the PreimageCache and
+	// forwarding the HTLC (committing the circuit), we aquire a multimutex
+	// for this paymentID.
+	s.paymentIDMtx.Lock(paymentID)
+
 	// If we already know the preimage, we can return it immediately.
 	p, ok := s.cfg.PreimageCache.LookupPreimage(htlc.PaymentHash[:])
 	if ok {
+		s.paymentIDMtx.Unlock(paymentID)
```

Comment:
Is it possible that anyone else locks in between this unlock and the lock above with the same `paymentID`? This `paymentID` is just generated and not known anywhere else at this point?

Reply:
Okay, I get this one too now. Generate of the `paymentID` is moved to router in the follow-up PR.

Reply:
Hmm.. but even then the paymentID passed into `SendHTLC` will always be unique?

Reply:
Added the commit that moves paymentID generation to the router to this PR, good idea :)